Add `PrefectDbtRunner` docs (#17055)
PrefectHQ/prefect: docs/integrations/prefect-dbt/index.mdx

This enables you to store sensitive credentials securely using Prefect blocks, and configure different targets based on the Prefect workspace.

For example, a Prefect variable called `target` can have a different value in development and production workspaces - `"dev"` and `"prod"`, respectively.
This allows you to use the same `profiles.yml` file for both environments, automatically referencing Snowflake blocks when running in production.

```yaml
example:
  outputs:
    dev:
      type: duckdb
      path: dev.duckdb
      threads: 1

    prod:
      type: snowflake
      account: "{{ prefect.blocks.snowflake-credentials.warehouse-access.account }}"
      user: "{{ prefect.blocks.snowflake-credentials.warehouse-access.user }}"
      password: "{{ prefect.blocks.snowflake-credentials.warehouse-access.password }}"
      database: "{{ prefect.blocks.snowflake-connector.prod-connector.database }}"
      schema: "{{ prefect.blocks.snowflake-connector.prod-connector.schema }}"
      warehouse: "{{ prefect.blocks.snowflake-connector.prod-connector.warehouse }}"
      threads: 4

  target: "{{ prefect.variables.target }}"
```

#### Logging

> The `PrefectDbtRunner` maps all dbt log levels to standard Python logging levels, so filtering for log levels like `WARNING` or `ERROR` in the Prefect UI applies to dbt's logs.

Reviewer: If you don't want to add "class" to all of these, you could alternatively say "`PrefectDbtRunner` maps all…".

```suggestion
The `PrefectDbtRunner` class maps all dbt log levels to standard Python logging levels, so filtering for log levels like `WARNING` or `ERROR` in the Prefect UI applies to dbt's logs.
```

> It also selects the appropriate logging target based on the invocation context: logging to a Prefect run logger when executed within a flow or task, and logging directly to the terminal when run outside a flow or task.

Reviewer: ```suggestion
It also selects the appropriate logging target based on the invocation context: logging to a Prefect run logger when executed within a flow or task, and logging directly to the terminal when run outside of a flow or task.
```

#### Failure handling

> The `PrefectDbtRunner` raises exceptions with standardized messages containing detailed information about the failure.

Reviewer: ```suggestion
The `PrefectDbtRunner` class raises exceptions with standardized messages containing detailed information about the failure.
```

```
Failures detected during invocation of dbt command 'build':
Test not_null_my_first_dbt_model_id failed with message: "Got 1 result, configured to fail if != 0"
```

It also offers a `raise_on_failure` option to control whether non-exception dbt failures, like failed tests, are raised as exceptions.

```python
from prefect import flow
from prefect_dbt import PrefectDbtRunner

@flow
def run_dbt():
    PrefectDbtRunner(
        raise_on_failure=False  # Failed tests will not fail the flow run
    ).invoke(["build"])

if __name__ == "__main__":
    run_dbt()
```

#### Events

> The `PrefectDbtRunner` automatically emits prefect events for completed dbt nodes with status information.

Reviewer: You have Prefect capitalized elsewhere in this doc (e.g. "Prefect runner")

```suggestion
The `PrefectDbtRunner` automatically emits Prefect events for completed dbt nodes with status information.
```

  "occurred": "2025-02-06T17:05:02.614Z",
  "payload": {
    "node_info": {
      "node_path": "test/my_second_dbt_model.sql",
      "node_name": "my_second_dbt_model",
      "unique_id": "model.test.my_second_dbt_model",
      "resource_type": "model",
      "materialized": "view",
      "node_status": "skipped",
      "node_started_at": "2025-02-06T17:05:02.614034",
      "meta": {},
      "node_relation": {
        "database": "dev",
        "schema": "main",
        "alias": "my_second_dbt_model",
        "relation_name": "\"dev\".\"main\".\"my_second_dbt_model\""
      }
    },
    "run_result": {
      "status": "skipped",
      "thread": "Thread-2 (worker)",
      "adapter_response": {}
    }
  },
  "received": "2025-02-06T17:05:02.843Z",
  "resource": {
    "prefect.resource.id": "dbt.model.test.my_second_dbt_model",
    "prefect.resource.name": "my_second_dbt_model",
    "dbt.node.status": "skipped"
  },
  "workspace": "87a4870c-8152-46cb-93aa-06482f9aa1df"
}
```

These events contain both the `node_info` and `run_result` for processed nodes in your dbt project, as well as the final status of the node's execution in the primary resource.

> This enables you to set up alerting for specific node states through Prefect's automations system. To be notified when this model is skipped, the following automation trigger could be set up:

Reviewer: Ventilated prose

```suggestion
This enables you to set up alerting for specific node states through Prefect's automations system.
To be notified when this model is skipped, the following automation trigger could be set up:
```

```json
{
  "type": "event",
  "match": {
    "prefect.resource.id": [
      "dbt.model.test.my_second_dbt_model"
    ],
    "dbt.node.status": "skipped"
  },
  "match_related": {},
  "after": [],
  "expect": [
    "my_second_dbt_model skipped"
  ],
  "for_each": [],
  "posture": "Reactive",
  "threshold": 1,
  "within": 0
}
```

#### Resources and Lineage

> The `PrefectDbtRunner` automatically emits lineage events describing the relationships between the resources in your dbt project.

Reviewer: ```suggestion
The `PrefectDbtRunner` class automatically emits lineage events describing the relationships between the resources in your dbt project.
```

> Resources can be viewed in Prefect Cloud on the flow run page. Clicking "View graph" will display a lineage graph with nodes directly upstream and downstream of the selected resource.

Reviewer: Ventilated prose. And prefer active tense to passive tense when it simplifies the content.

```suggestion
Resources can be viewed in Prefect Cloud on the flow run page.
Click "View graph" to display a lineage graph with nodes directly upstream and downstream of the selected resource.
```

Author: noted and noted

## dbt Core

### prefect-dbt 0.7.0 and later

<Info>
During the prerelease phase of `prefect-dbt==0.7.0`, use the `--pre` flag to install the latest release candidate.

```bash
pip install prefect-dbt --pre
```
</Info>

> Versions 0.7.0 and later of `prefect-dbt` include the `PrefectDbtRunner` class, which provides an improved interface for running dbt Core commands with better logging, failure handling, and additional features.

Reviewer: I think this header is fine. The table of contents in the right column shows that this section is followed by an equivalent section for prefect-dat <0.7.0, so I don't think people will be confused.

Resources can be viewed in Prefect Cloud on the flow run page. Clicking "View graph" will display a lineage graph with nodes directly upstream and downstream of the selected resource.

When running in Prefect Cloud, the lineage graph for executed nodes will be visible on the flow run page, showing upstream and downstream dependencies.

To describe resource that exist outside of your dbt project and are upstream of a given node, supply the resource in your dbt config.
For example, to describe a BigQuery table as an upstream resource of a dbt model, add the following to your model's config:

```yaml
prefect:
  upstream_resources:
    - name: "my_bigquery_table"
      role: "table"
      id: "bigquery://my-project.my_dataset.my_table"
```

#### Native dbt configuration

You can disable these features for all resources in your dbt project config, or for specific resources in their own config:

```yaml
prefect:
  emit_events: False  # Disable all events
  emit_node_events: False  # Disable node events
  emit_lineage_events: False  # Disable lineage events
```

> ### prefect-dbt < 0.7.0

Reviewer: Consider rephrasing this section header to match the format of the 0.7.0 section header.

```suggestion
### prefect-dbt 0.6.6 and earlier
```
